File: scripts/generate-feeds.js
import PocketBase from 'pocketbase';
import { Feed } from 'feed';
import ical from 'ical-generator';
import fs from 'fs';
import path from 'path';

const pb = new PocketBase('https://data.suomenpalikkayhteiso.fi');

const dayAbbr = ['su', 'ma', 'ti', 'ke', 'to', 'pe', 'la'];

function formatDateInHelsinki(utcString, allDay = false) {
	const date = new Date(utcString.endsWith('Z') ? utcString : utcString + 'Z');
	const helsinkiDate = new Date(date.toLocaleString('en-US', { timeZone: 'Europe/Helsinki' }));

	const day = String(helsinkiDate.getDate());
	const month = String(helsinkiDate.getMonth() + 1);
	const dateStr = `${day}.${month}.`;

	if (allDay) {
		return dateStr;
	} else {
		const hours = String(helsinkiDate.getHours()).padStart(2, '0');
		const minutes = String(helsinkiDate.getMinutes()).padStart(2, '0');
		const timeStr = `${hours}.${minutes}`;
		return `${dateStr} ${timeStr}`;
	}
}

async function generateFeeds() {
	const events = await pb.collection('events').getFullList({
		sort: 'start_date',
		filter: 'state = "published"'
	});

	// Create images directory if it doesn't exist
	const imagesDir = path.join('static', 'images');
	if (!fs.existsSync(imagesDir)) {
		fs.mkdirSync(imagesDir, { recursive: true });
	}

	// Download and cache event images
	const imageUrls = new Map();
	for (const event of events) {
		if (event.image) {
			try {
				const imageUrl = `https://data.suomenpalikkayhteiso.fi/api/files/events/${event.id}/${event.image}`;
				const response = await fetch(imageUrl);
				if (response.ok) {
					const buffer = await response.arrayBuffer();
					const localImagePath = path.join(imagesDir, `${event.id}_${event.image}`);
					fs.writeFileSync(localImagePath, Buffer.from(buffer));
					imageUrls.set(event.id, `/images/${event.id}_${event.image}`);
				}
			} catch (error) {
				console.warn(`Failed to download image for event ${event.id}:`, error.message);
			}
		}
	}

	// Generate RSS feed
	const feed = new Feed({
		title: 'Palikkakalenteri',
		description: 'Suomen Palikkayhteisö ry:n Palikkakalenteri',
		id: 'https://kalenteri.suomenpalikkayhteiso.fi/',
		link: 'https://kalenteri.suomenpalikkayhteiso.fi/',
		language: 'fi',
		image: 'https://kalenteri.suomenpalikkayhteiso.fi/logo.png',
		favicon: 'https://kalenteri.suomenpalikkayhteiso.fi/favicon.ico',
		copyright: 'Suomen Palikkayhteisö ry',
		updated: new Date(),
		generator: 'Emmet',
		feedLinks: {
			rss: 'https://kalenteri.suomenpalikkayhteiso.fi/kalenteri.rss',
			atom: 'https://kalenteri.suomenpalikkayhteiso.fi/kalenteri.atom'
		},
		author: {
			name: 'Suomen Palikkayhteisö ry',
			email: 'suomenpalikkayhteisory@outlook.com',
			link: 'https://suomenpalikkayhteiso.fi/'
		}
	});

	events.forEach((event) => {
		const startFormatted = formatDateInHelsinki(event.start_date, event.all_day);
		let dateStr = startFormatted;
		if (event.end_date) {
			const endFormatted = formatDateInHelsinki(event.end_date, event.all_day);
			if (startFormatted !== endFormatted) {
				if (event.all_day) {
					const startParts = startFormatted.split('.');
					const endParts = endFormatted.split('.');
					if (startParts[1] === endParts[1]) {
						// Same month, drop month from start
						dateStr = `${startParts[0]}.\u2013${endParts[0]}.${startParts[1]}.`;
					} else {
						dateStr += `\u2013${endFormatted}`;
					}
				} else {
					// For timed events, if same date, show only end time
					const startDatePart = startFormatted.split(' ')[0];
					const endDatePart = endFormatted.split(' ')[0];
					if (startDatePart === endDatePart) {
						const endTime = endFormatted.split(' ')[1];
						dateStr += `\u2013${endTime}`;
					} else {
						dateStr += `\u2013${endFormatted}`;
					}
				}
			}
		}
		if (event.all_day) {
			const startDate = new Date(
				event.start_date.endsWith('Z') ? event.start_date : event.start_date + 'Z'
			);
			const startHelsinki = new Date(
				startDate.toLocaleString('en-US', { timeZone: 'Europe/Helsinki' })
			);
			const startDay = startHelsinki.getDay();
			let prefix = '';
			if (event.end_date && event.end_date !== event.start_date) {
				const endDate = new Date(
					event.end_date.endsWith('Z') ? event.end_date : event.end_date + 'Z'
				);
				const endHelsinki = new Date(
					endDate.toLocaleString('en-US', { timeZone: 'Europe/Helsinki' })
				);
				const endDay = endHelsinki.getDay();
				const duration = Math.ceil((endDate - startDate) / (24 * 60 * 60 * 1000)) + 1;
				if (duration < 8) {
					prefix = `${dayAbbr[startDay]}\u2013${dayAbbr[endDay]} `;
				}
			} else {
				prefix = `${dayAbbr[startDay]} `;
			}
			dateStr = prefix + dateStr;
		} else {
			// For timed events, add day abbr and "klo" for single day
			const startDate = new Date(
				event.start_date.endsWith('Z') ? event.start_date : event.start_date + 'Z'
			);
			const startHelsinki = new Date(
				startDate.toLocaleString('en-US', { timeZone: 'Europe/Helsinki' })
			);
			const startDay = startHelsinki.getDay();
			const isSingleDay =
				!event.end_date ||
				formatDateInHelsinki(event.start_date, false).split(' ')[0] ===
					formatDateInHelsinki(event.end_date, false).split(' ')[0];
			if (isSingleDay) {
				const parts = dateStr.split(' ');
				if (parts.length === 2) {
					dateStr = `${dayAbbr[startDay]} ${parts[0]} klo ${parts[1]}`;
				}
			}
		}
		const description = event.description || event.title;
		const content = description + '\n\n' + dateStr;
		const eventUrl = event.url || `https://kalenteri.suomenpalikkayhteiso.fi/#/events/${event.id}`;

		const feedItem = {
			title: `${event.all_day ? dateStr : formatDateInHelsinki(event.start_date, true)} ${event.title} | ${event.location}`,
			id: event.id,
			link: eventUrl,
			description: content,
			date: new Date(event.updated),
			published: new Date(event.created),
			author: [
				{
					name: 'Suomen Palikkayhteisö ry'
				}
			]
		};

		// Add image if available
		const imageUrl = imageUrls.get(event.id);
		if (imageUrl) {
			feedItem.image = `https://kalenteri.suomenpalikkayhteiso.fi${imageUrl}`;
		}

		feed.addItem(feedItem);
	});

	fs.writeFileSync('static/kalenteri.rss', feed.rss2());
	fs.writeFileSync('static/kalenteri.atom', feed.atom1());
	if (fs.existsSync('build')) {
		fs.writeFileSync('build/kalenteri.rss', feed.rss2());
		fs.writeFileSync('build/kalenteri.atom', feed.atom1());
	}

	// Generate ICAL feed
	const calendar = ical({
		title: 'Palikkakalenteri',
		description: 'Suomen Palikkayhteisö ry:n Palikkakalenteri',
		timezone: 'Europe/Helsinki'
	});

	events.forEach((event) => {
		const startDate = new Date(event.start_date + (event.start_date.includes('Z') ? '' : 'Z')); // Ensure UTC
		const endDate = event.end_date
			? new Date(event.end_date + (event.end_date.includes('Z') ? '' : 'Z')) // Ensure UTC
			: new Date(startDate.getTime() + (event.all_day ? 24 * 60 * 60 * 1000 : 60 * 60 * 1000)); // Default 1 day for all_day, 1 hour otherwise

		const description = event.description || event.title;
		const eventUrl = event.url || `https://kalenteri.suomenpalikkayhteiso.fi/#/events/${event.id}`;

		const eventData = {
			id: event.id,
			start: startDate,
			end: endDate,
			summary: `${event.title} | ${event.location}`,
			description: description,
			url: eventUrl,
			timezone: 'Europe/Helsinki'
		};

		if (event.location) {
			eventData.location = event.location;
		}

		if (event.all_day) {
			eventData.allDay = true;
		}

		calendar.createEvent(eventData);
	});

	fs.writeFileSync('static/kalenteri.ics', calendar.toString());
	if (fs.existsSync('build')) {
		fs.writeFileSync('build/kalenteri.ics', calendar.toString());
	}
}

generateFeeds();

File: scripts/generate_feeds.py
#!/usr/bin/env python3
"""
Feed generation (RSS, Atom, JSON) for the event calendar
"""

import os
import base64
from datetime import datetime, timezone
from pathlib import Path
from icalendar import Calendar, Event as ICalEvent
from feedgen.feed import FeedGenerator
from generate_utils import *

BASE_URL = 'https://kalenteri.suomenpalikkayhteiso.fi'

def download_event_images(events):
    """Download event images and return mapping of event IDs to local paths"""
    image_urls = {}

    for event in events:
        if not event.get('image'):
            continue

        try:
            image_url = f"{POCKETBASE_URL}/api/files/events/{event['id']}/{event['image']}"
            response = requests.get(image_url)
            response.raise_for_status()

            relative_path = f"images/{event['id']}_{event['image']}"
            write_static_file(relative_path, response.content)
            image_urls[event['id']] = f"/images/{event['id']}_{event['image']}"

        except Exception as e:
            print(f"Failed to download image for event {event['id']}: {e}")

    return image_urls

def generate_feeds(events):
    """Generate RSS, Atom, and JSON feeds"""
    # image_urls = download_event_images(events)
    image_urls = {}  # Skip image downloading for now

    # Create individual ICS files and data URIs
    event_ics_data_uris = {}
    event_ics_contents = {}

    for event in events:
        calendar = Calendar()
        calendar.add('title', 'Palikkakalenteri')
        calendar.add('description', 'Suomen Palikkayhteisö ry:n Palikkakalenteri')
        calendar.add('timezone', 'Europe/Helsinki')

        start_date = to_helsinki_date(event['start_date'])
        end_date = to_helsinki_date(event.get('end_date', event['start_date']))

        if event.get('all_day', False):
            end_date = end_date.replace(hour=23, minute=59, second=59)
        else:
            if not event.get('end_date') or event['end_date'] == event['start_date']:
                end_date = start_date.replace(hour=start_date.hour + 1)

        description = event.get('description', event['title'])
        event_url = event.get('url')

        ical_event = ICalEvent()
        ical_event.add('id', f"{BASE_URL}/#/events/{event['id']}")
        ical_event.add('start', start_date)
        ical_event.add('end', end_date)
        ical_event.add('summary', event['title'])
        ical_event.add('description', description)
        ical_event.add('timezone', 'Europe/Helsinki')

        if event_url:
            ical_event.add('url', event_url)

        if event.get('location'):
            ical_event.add('location', event['location'])
            if event.get('point') and event['point'].get('lat') and event['point'].get('lon'):
                ical_event.add('geo', (event['point']['lat'], event['point']['lon']))

        if event.get('all_day', False):
            ical_event.add('all_day', True)

        calendar.add_component(ical_event)

        ics_content = calendar.to_ical().decode('utf-8')
        write_static_file(f"events/{event['id']}.ics", ics_content)

        event_ics_contents[event['id']] = ics_content

        # Create data URI for enclosure
        ics_base64 = base64.b64encode(ics_content.encode('utf-8')).decode('utf-8')
        data_uri = f"data:text/calendar;charset=utf-8;base64,{ics_base64}"
        event_ics_data_uris[event['id']] = data_uri

        # Generate HTML file for the event
        event_html = f'''<!DOCTYPE html>
<html lang="fi">
<head>
<meta charset="UTF-8">
<title>{event['title']} - Palikkakalenteri</title>
<meta name="viewport" content="width=device-width, initial-scale=1">
<style>
body {{ font-family: Arial, sans-serif; margin: 20px; text-align: center; }}
h1 {{ color: #333; }}
p {{ margin: 20px 0; }}
a {{ display: inline-block; padding: 12px 24px; background-color: #007bff; color: white; text-decoration: none; border-radius: 4px; }}
a:hover {{ background-color: #0056b3; }}
</style>
</head>
<body>
<h1>{event['title']}</h1>
<p>{event.get('location', '') and f"{event['location']}<br>" or ''}{format_event_display_date(event)}</p>
<p>{event.get('description', '')}</p>
<p><a href="{event['id']}.ics">Lisää kalenteriin</a></p>
</body>
</html>'''
        write_static_file(f"events/{event['id']}.html", event_html)

    # Generate feeds
    fg = FeedGenerator()
    fg.title('Palikkakalenteri')
    fg.description('Suomen Palikkayhteisö ry:n Palikkakalenteri')
    fg.id(BASE_URL)
    fg.link(href=BASE_URL, rel='alternate')
    fg.logo(f"{BASE_URL}/logo.png")
    fg.icon(f"{BASE_URL}/favicon.ico")
    fg.copyright('Suomen Palikkayhteisö ry')
    fg.updated(datetime.now(timezone.utc))

    fg.link(href=f"{BASE_URL}/kalenteri.rss", rel='self', type='application/rss+xml')
    fg.link(href=f"{BASE_URL}/kalenteri.atom", rel='alternate', type='application/atom+xml')

    for event in events:
        date_str = format_event_display_date(event)
        description = event.get('description', event['title'])
        content = f"{description}\n\n{date_str}"
        event_url = event.get('url')

        fe = fg.add_entry()
        fe.title(f"{date_str} {event['title']}" + (event.get('location') and f" | {event['location']}" or ''))
        fe.id(f"{BASE_URL}/#/events/{event['id']}")
        if event_url:
            fe.link(href=event_url, rel='alternate')
        fe.description(content)
        fe.updated(datetime.fromisoformat(event['updated'][:-1]).replace(tzinfo=timezone.utc))
        fe.published(datetime.fromisoformat(event['created'][:-1]).replace(tzinfo=timezone.utc))

        fe.author(name='Suomen Palikkayhteisö ry', email='suomenpalikkayhteisory@outlook.com')

        image_url = image_urls.get(event['id'])
        if image_url:
            fe.enclosure(f"{BASE_URL}{image_url}", 0, 'image/jpeg')  # Size will be determined by file

        # Add ICS enclosure
        ics_data_uri = event_ics_data_uris.get(event['id'])
        if ics_data_uri:
            ics_file_path = Path('static') / f"events/{event['id']}.ics"
            length = ics_file_path.stat().st_size if ics_file_path.exists() else 0
            fe.enclosure(f"{BASE_URL}/events/{event['id']}.ics", length, 'text/calendar')

    # Generate RSS
    rss = fg.rss_str(pretty=True)
    write_static_file('kalenteri.rss', rss.decode('utf-8'))

    # Generate Atom
    atom = fg.atom_str(pretty=True)
    # Postprocess Atom feed to fix incorrect enclosure type for ICS files
    atom_str = atom.decode('utf-8').replace('type="image/ics"', 'type="text/calendar"')
    write_static_file('kalenteri.atom', atom_str)

    # Generate JSON Feed (TODO: implement JSON feed generation)
    # json_feed = fg.json_feed_str(pretty=True)
    # write_static_file('kalenteri.json', json_feed.decode('utf-8'))
    print('Note: JSON feed generation not yet implemented')

    print(f'Generated feeds and individual event files')

if __name__ == '__main__':
    # Test the feed generation
    pb_url = create_pocketbase_client()
    events = fetch_published_events(pb_url)
    generate_feeds(events)